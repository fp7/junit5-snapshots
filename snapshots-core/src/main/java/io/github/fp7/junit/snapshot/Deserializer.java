package io.github.fp7.junit.snapshot;

import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.junit.jupiter.api.Assertions;

final class Deserializer {

  static Map<String, List<String>> read(Path path) {
    try {
      Map<String, List<String>> result = new HashMap<>();
      var strings = Files.readAllLines(path).stream().skip(2).collect(Collectors.toList());

      StringWriter stringHolder = null;
      PrintWriter w = null;
      String name = null;
      var started = false;

      for (var string : strings) {
        if (string.startsWith("exports[`") && string.endsWith("`] = `")) {
          stringHolder = new StringWriter();
          w = new PrintWriter(stringHolder);
          name = string.substring("exports[`".length(), string.length() - "`] = `".length());
          name = name.substring(0, name.lastIndexOf(" "));
          started = true;

        } else if (string.equals("`;")) {
          Assertions.assertNotNull(stringHolder);
          Assertions.assertNotNull(name);
          started = false;

          String snapshot = stringHolder.toString();
          snapshot = snapshot.substring(0, snapshot.length() - 1);
          result.merge(
              name,
              List.of(snapshot),
              (r1, r2) -> Stream.concat(r1.stream(), r2.stream()).collect(Collectors.toList()));

        } else if (started) {
          Assertions.assertNotNull(w);
          w.println(string);
        }
      }

      return result;
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }
}
