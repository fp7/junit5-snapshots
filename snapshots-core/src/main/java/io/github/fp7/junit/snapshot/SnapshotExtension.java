package io.github.fp7.junit.snapshot;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Objects;
import java.util.Optional;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
import org.junit.jupiter.api.extension.BeforeTestExecutionCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
import org.junit.jupiter.api.extension.ParameterContext;
import org.junit.jupiter.api.extension.ParameterResolutionException;
import org.junit.jupiter.api.extension.ParameterResolver;

public class SnapshotExtension implements BeforeTestExecutionCallback, AfterTestExecutionCallback,
    ParameterResolver {

  private static final Namespace ns = Namespace.create("snapshots-data");

  private static Path snapshotPath;

  public static Path getSnapshotPath() {
    return Objects.requireNonNull(snapshotPath);
  }

  public SnapshotExtension(String foo) {
  }

  private String getName(ExtensionContext ctx) {
    var l = new ArrayList<String>();
    Optional<ExtensionContext> tmp = Optional.of(ctx);
    while (tmp.isPresent()) {
      l.add(tmp.get().getDisplayName());
      tmp = tmp.get().getParent();
    }
    Collections.reverse(l);

    return String.join(" ", l);



  }

  @Override
  public void beforeTestExecution(ExtensionContext context) throws Exception {

    context.getStore(ns).put("cur", getName(context));

  }

  @Override
  public void afterTestExecution(ExtensionContext context) throws Exception {
    System.out.println("context.getStore(ns).get(\"cur\") = " + context.getStore(ns).get("cur"));

    snapshotPath = null;
  }



  @Override
  public boolean supportsParameter(ParameterContext parameterContext,
      ExtensionContext extensionContext) throws ParameterResolutionException {
    return parameterContext.getParameter().getType().equals(Snapshot.class);
  }

  @Override
  public Object resolveParameter(ParameterContext parameterContext,
      ExtensionContext extensionContext) throws ParameterResolutionException {

    return new Snapshot("foo");
  }

  public static final class Snapshot{
    private final String snapshot;

    public Snapshot(String snapshot) {
      this.snapshot = snapshot;
    }

    public void matches(Object target){
      Assertions.assertEquals(snapshot, target);

    }
  }
}
