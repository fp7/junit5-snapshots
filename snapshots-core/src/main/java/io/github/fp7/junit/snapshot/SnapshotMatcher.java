package io.github.fp7.junit.snapshot;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

public class SnapshotMatcher {

  private final String testName;

  private final Snapshots snapshots;

  public SnapshotMatcher(String testName, Snapshots snapshots) {
    this.testName = testName;
    this.snapshots = snapshots;
  }

  public void matchesSnapshot(Object target){
    snapshots.matches(testName, new GsonBuilder().setPrettyPrinting().create().toJson(target));
  }
}
