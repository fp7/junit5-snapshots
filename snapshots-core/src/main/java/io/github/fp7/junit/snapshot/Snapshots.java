package io.github.fp7.junit.snapshot;

import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Assertions;

public final class Snapshots {

  private final Path snapshot;

  private final Map<String, List<String>> testSnapshots = new HashMap<>();

  Snapshots(Path snapshot) {
    this.snapshot = snapshot;
  }

  void loadSnapshots() {
    System.out.println("loading snapshot");
  }

  public void matches(Object target) {
    Assertions.assertEquals(snapshot, target);
  }

  public void finishTest() {
    System.out.println("Finishing snapshots");
  }
}
