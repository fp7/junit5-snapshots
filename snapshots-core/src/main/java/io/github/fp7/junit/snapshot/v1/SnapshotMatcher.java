package io.github.fp7.junit.snapshot.v1;

import com.google.gson.GsonBuilder;

public class SnapshotMatcher {

  private final String testName;

  private final Snapshots snapshots;

  public SnapshotMatcher(String testName, Snapshots snapshots) {
    this.testName = testName;
    this.snapshots = snapshots;
  }

  public void matchesSnapshot(Object target) {
    snapshots.matches(testName, new GsonBuilder().setPrettyPrinting().create().toJson(target));
  }
}
