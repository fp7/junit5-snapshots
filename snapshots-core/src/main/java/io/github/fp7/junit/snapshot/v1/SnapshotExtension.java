/*********************************************************************
* Copyright (c) 2019 Finn Petersen
*
* This program and the accompanying materials are made
* available under the terms of the Eclipse Public License 2.0
* which is available at https://www.eclipse.org/legal/epl-2.0/
*
* SPDX-License-Identifier: EPL-2.0
**********************************************************************/

package io.github.fp7.junit.snapshot.v1;

import io.github.fp7.junit.snapshot.v1.Snapshots.SnapshotRecorder;
import io.github.fp7.junit.snapshot.v1.Snapshots.SnapshotVerifier;
import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
import org.junit.jupiter.api.extension.BeforeAllCallback;
import org.junit.jupiter.api.extension.BeforeTestExecutionCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
import org.junit.jupiter.api.extension.ParameterContext;
import org.junit.jupiter.api.extension.ParameterResolutionException;
import org.junit.jupiter.api.extension.ParameterResolver;

public class SnapshotExtension
    implements BeforeAllCallback,
        BeforeTestExecutionCallback,
        AfterTestExecutionCallback,
        AfterAllCallback,
        ParameterResolver {

  private static final Namespace SNAPSHOTS = Namespace.create("SNAPSHOTS-data");

  public SnapshotExtension() {}

  @Override
  public void beforeTestExecution(ExtensionContext context) throws Exception {}

  @Override
  public void afterTestExecution(ExtensionContext context) throws Exception {}

  @Override
  public boolean supportsParameter(
      ParameterContext parameterContext, ExtensionContext extensionContext)
      throws ParameterResolutionException {
    return parameterContext.getParameter().getType().equals(SnapshotMatcher.class);
  }

  @Override
  public Object resolveParameter(
      ParameterContext parameterContext, ExtensionContext extensionContext)
      throws ParameterResolutionException {
    return new SnapshotMatcher(
        testName(extensionContext),
        extensionContext.getStore(SNAPSHOTS).get("current", Snapshots.class));
  }

  private String testName(ExtensionContext extensionContext) {
    return String.format(
        "%s#%s",
        extensionContext.getRequiredTestClass().getName(), extensionContext.getDisplayName());
  }

  @Override
  public void afterAll(ExtensionContext context) throws Exception {
    if (context.getExecutionException().isPresent()) {
      return;
    }
    context.getStore(SNAPSHOTS).get("current", Snapshots.class).finishTest();
  }

  @Override
  public void beforeAll(ExtensionContext context) throws Exception {
    Path file = snapshotFile(context);
    Snapshots snapshots;
    if (Files.exists(file)) {
      snapshots = new SnapshotVerifier(file);
      snapshots.loadSnapshots();
    } else {
      snapshots = new SnapshotRecorder(file);
    }

    context.getStore(SNAPSHOTS).put("current", snapshots);
  }

  private Path snapshotFile(ExtensionContext ctx) {
    return Path.of(
        "src/test/java",
        String.format(
            "%s.java.snap", ctx.getRequiredTestClass().getName().replace(".", File.separator)));
  }
}
