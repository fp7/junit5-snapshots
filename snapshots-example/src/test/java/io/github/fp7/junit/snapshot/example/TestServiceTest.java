package io.github.fp7.junit.snapshot.example;

import io.github.fp7.junit.snapshot.SnapshotExtension;
import io.github.fp7.junit.snapshot.SnapshotMatcher;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(SnapshotExtension.class)
class TestServiceTest {

  @Test
  void checkAgainstFoo(SnapshotMatcher s) {
    s.matchesSnapshot("foo");
  }

  @Test
  void checkBacktickSemicolon(SnapshotMatcher s){
    s.matchesSnapshot("`;");
  }

  @Test
  void checkEmptyLines(SnapshotMatcher s){
    s.matchesSnapshot("hallo \n\nworld!");
  }
}
