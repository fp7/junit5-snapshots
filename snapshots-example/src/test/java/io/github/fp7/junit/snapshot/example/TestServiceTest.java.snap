// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`io.github.fp7.junit.snapshot.example.TestServiceTest#checkBacktickSemicolon(SnapshotMatcher) 1`] = `
\`;
`;

exports[`io.github.fp7.junit.snapshot.example.TestServiceTest#checkEmptyLines(SnapshotMatcher) 1`] = `
hallo 

world!
`;

exports[`io.github.fp7.junit.snapshot.example.TestServiceTest#checkAgainstFoo(SnapshotMatcher) 1`] = `
foo
`;
